File: slSig/Impressao/docs/XPSCardPrinterSDK74099/samples/cpp/smartcard_singlewire_mifare/MiFare_Command.cpp
////////////////////////////////////////////////////////////////////////////////
// Copyright (c) Datacard Corporation.  All Rights Reserved.
//
// construct mifare commands for duali reader.
//
// see "CCID_Protocol_spec_120224_.pdf"
////////////////////////////////////////////////////////////////////////////////
#include "StdAfx.h"
#include "Mifare_Common.h"
#include "MiFare_Command.h"

using namespace dxp01sdk;

std::vector <byte> MiFare_Command::CreateGetCardStatusCommand()
{
    // see pg 23, "CCID_Protocol_spec_120224_.pdf"

    const byte commandBytes[] = {0xFE, 0x17, 0xFE, 0xFE, 0x00};
    const std::vector <byte> getCardStatusCommand(commandBytes, commandBytes + ARRAYSIZE(commandBytes));
    return getCardStatusCommand;
}

std::vector <byte> MiFare_Command::CreateLoadKeysCommand(
    const KeyType       keyType,
    const byte          sector,
    std::vector <byte>  keyBytes)
{
    // see pg 11, "CCID_Protocol_spec_120224_.pdf"

    if (0 == keyBytes.size()) {
        throw std::runtime_error("keyBytes has no data");
    }

    MiFare_Common miFareCommon;
    miFareCommon.CheckSectorNumber(sector);

    byte commandBytes[] = {0xFD, 0x2F, (byte) keyType, sector, 0x06};
    std::vector <byte> loadKeysCommand(
        commandBytes,
        commandBytes + ARRAYSIZE(commandBytes));

    // append keyBytes onto the end of the commandBytes:
    loadKeysCommand.insert(loadKeysCommand.end(), keyBytes.begin(), keyBytes.end());

    return loadKeysCommand;
}

std::vector <byte> MiFare_Command::CreateReadBlockCommand(
    const KeyType   keyType,
    const byte      sector,
    const byte      block)
{
    // see pg 12, "CCID_Protocol_spec_120224_.pdf"

    MiFare_Common miFareCommon;
    miFareCommon.CheckSectorNumber(sector);
    miFareCommon.CheckBlockNumber(block);

    const byte commandBytes[] = {0xFD, 0x35, (byte) keyType, sector, 0x01, block};
    const std::vector <byte> readBlockCommand(
        commandBytes,
        commandBytes + ARRAYSIZE(commandBytes));
    return readBlockCommand;
}

std::vector <byte> MiFare_Command::CreateWriteBlockCommand(
    const KeyType       keyType,
    const byte          sector,
    const byte          block,
    std::vector <byte>  blockData)
{
    // see pg 13, "CCID_Protocol_spec_120224_.pdf"

    if (0 == blockData.size()) {
        throw std::runtime_error("blockData has no data");
    }

    MiFare_Common miFareCommon;
    miFareCommon.CheckSectorNumber(sector);
    miFareCommon.CheckBlockNumber(block);

    const byte commandBytes[] = {0xFD, 0x37, (byte) keyType, sector, 0x11, block};
    std::vector <byte> writeBlockCommand(
        commandBytes,
        commandBytes + ARRAYSIZE(commandBytes));

    // append blockData onto the end of the commandBytes:
    writeBlockCommand.insert(writeBlockCommand.end(), blockData.begin(), blockData.end());

    return writeBlockCommand;
}

File: slSig/Impressao/docs/XPSCardPrinterSDK74099/samples/csharp/smartcard_singlewire_mifare/Mifare_Command.cs
////////////////////////////////////////////////////////////////////////////////
// Copyright (c) Datacard Corporation.  All Rights Reserved.
//
// construct mifare commands for duali reader.
//
// see "CCID_Protocol_spec_120224_.pdf"
////////////////////////////////////////////////////////////////////////////////
using System.Collections.Generic;
using System.Diagnostics;

namespace dxp01sdk.mifare {

    public class Mifare_Command {

        public enum KeyType : byte { A = 0, B = 4 }

        // see pg 23, "CCID_Protocol_spec_120224_.pdf"
        public byte[] CreateGetCardStatusCommand() {
            var commandBytes = new byte[] { 0xFE, 0x17, 0xFE, 0xFE, 0x00 };
            return commandBytes;
        }

        // see pg 11, "CCID_Protocol_spec_120224_.pdf"
        public byte[] CreateLoadKeysCommand(KeyType keyType, byte sector, byte[] keyBytes) {
            Mifare_Common.CheckSectorNumber(sector);
            Debug.Assert(keyBytes.Length == 6);
            var commandBytes = new List<byte>(new byte[] { 0xFD, 0x2F, (byte) keyType, sector, 0x06 });
            commandBytes.AddRange(keyBytes);
            return commandBytes.ToArray();
        }

        // see pg 12, "CCID_Protocol_spec_120224_.pdf"
        public byte[] CreateReadBlockCommand(KeyType keyType, byte sector, byte block) {
            Mifare_Common.CheckSectorNumber(sector);
            Mifare_Common.CheckBlockNumber(block);
            var commandBytes = new byte[] { 0xFD, 0x35, (byte) keyType, sector, 0x01, block };
            return commandBytes;
        }

        // see pg 13, "CCID_Protocol_spec_120224_.pdf"
        public byte[] CreateWriteBlockCommand(KeyType keyType, byte sector, byte block, byte[] blockData) {
            Mifare_Common.CheckSectorNumber(sector);
            Mifare_Common.CheckBlockNumber(block);
            var commandBytes = new List<byte>(new byte[] { 0xFD, 0x37, (byte) keyType, sector, 0x11, block });
            commandBytes.AddRange(blockData);
            return commandBytes.ToArray();
        }
    }
}